### PR TITLE
[SPARK-49003][SQL] Fix interpreted code path hashing to be collation aware

### PR DESCRIPTION
**Pull request by ilicmarkodb**
### What changes were proposed in this pull request?
Changed hash function to be collation aware. This change is just for interpreted code path. Codegen hashing was already collation aware.

### Why are the changes needed?
We were getting the wrong hash for collated strings because our hash function only used the binary representation of the string. It didn't take collation into account.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests to `HashExpressionsSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 1]
```diff
@@ -620,6 +620,33 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkHiveHashForDecimal("123456.123456789012345678901234567890", 38, 31, 1728235666)
   }
 
+  for (collation <- Seq("UTF8_LCASE", "UNICODE_CI")) {
+    test(s"hash equality for collated $collation strings") {
+      val s1 = "aaa"
+      val s2 = "AAA"
+
+      // Interpreted hash value for s1
+      val hash = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
```
Comment (stefankandic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:629]:
```suggestion
      val interpretedHash = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 2]
```diff
@@ -620,6 +620,33 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+
+      // Check if interpreted hash value is same as codegen for s1
```
Comment (stefankandic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:631]: won't this check for both codegen and interpreted path for s1?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 3]
```diff
@@ -565,7 +565,15 @@ abstract class InterpretedHashFunction {
       case a: Array[Byte] =>
         hashUnsafeBytes(a, Platform.BYTE_ARRAY_OFFSET, a.length, seed)
       case s: UTF8String =>
-        hashUnsafeBytes(s.getBaseObject, s.getBaseOffset, s.numBytes(), seed)
+        dataType match {
```
Comment (stefankandic) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:568]: this code doesn't seem properly formatted

**Review by stefankandic — CHANGES_REQUESTED** [review]
Looks good overall, left some small comments. Also please change the title to be in format [SPARK-XYZU] Title, and update the description to be clear that you are only changing the interpreted code path and that the codegen is already collation aware

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 4]
```diff
@@ -620,6 +620,33 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+      checkEvaluation(Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42), hash)
+      // Check if s2's interpreted and codegen hash values are same as s1's
+      checkEvaluation(Murmur3Hash(Seq(Collate(Literal(s2), collation)), 42), hash)
+    }
+  }
+
+  for (collation <- Seq("UTF8_BINARY", "UNICODE")) {
```
Comment (stefankandic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:638]: could you maybe combine these tests and just check if the collation `supportsBinaryEquality` and do different checks based on that?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala (hunk 2 above)
Comment (ilicmarkodb) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:631]: yes, but my didn't want to write something like 'Check if interpreted hash value is same as interpreted and codegen for s1'

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 5]
```diff
@@ -620,6 +620,29 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
   }
 
+  for (collation <- Seq("UTF8_LCASE", "UNICODE_CI", "UTF8_BINARY")) {
+    test(s"hash equality for collated $collation strings") {
+      val s1 = "aaa"
+      val s2 = "AAA"
+
+      if (CollationFactory.fetchCollation(collation).supportsBinaryEquality) {
+        // Calculate interpreted hash values for s1 and s2
+        val interpretedHash1 = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
```
Comment (stefankandic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:630]: right now we are just checking this for the interpreted path, could we do something else instead like call `checkEvaluation` for both with the expected hash value?

**Comment by stefankandic** [comment]
please update the PR title

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 6]
```diff
@@ -25,13 +25,13 @@ import scala.language.implicitConversions
 
 import org.apache.commons.codec.digest.DigestUtils
 import org.scalatest.exceptions.TestFailedException
-
```
Comment (uros-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:28]: scalastyle fix

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 7]
```diff
@@ -620,6 +620,29 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
...
+        val interpretedHash1 = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
+        val interpretedHash2 = Murmur3Hash(Seq(Collate(Literal(s2), collation)), 42).eval()
+
+        assert(interpretedHash1 != interpretedHash2)
+      } else {
+        // Interpreted hash value for s1
+        val interpretedHash = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
+
+        // Check if interpreted hash value is same as codegen for s1
+        checkEvaluation(Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42), interpretedHash)
```
Comment (uros-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:639]: `checkEvaluation` will check both the interpreted path and gencode, so we should be able to just do:
```
    val hashExpr = Murmur3Hash(Seq(Collate(Literal(s1), collation)), 42).eval()
    checkEvaluation(hashExpr, expectedHash)
```

**Review by stefankandic — APPROVED** [review]
LGTM

**Comment by ilicmarkodb** [comment]
@cloud-fan  can you please take a look?

**Review by cloud-fan — APPROVED** [review]

**Comment by ilicmarkodb** [comment]
@mihailom-db can you please review change for your test?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
@ilicmarkodb can you do a rebase to trigger the CI?

**Review by uros-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala [hunk 8]
```diff
@@ -607,6 +607,8 @@ class CollationExpressionWalkerSuite extends SparkFunSuite with SharedSparkSessi
       // need to skip as plans differ in STRING <-> STRING COLLATE UTF8_LCASE
       "current_timezone",
       "schema_of_variant",
+      "hash",
+      "xxhash64",
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationExpressionWalkerSuite.scala:611]: nit: Is this the right group for these expressions? Don't we expect results to differ, as hashes are differnet?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala [hunk 9]
```diff
@@ -620,6 +620,30 @@ class HashExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkHiveHashForDecimal("123456.123456789012345678901234567890", 38, 31, 1728235666)
   }
 
+  for (collation <- Seq("UTF8_LCASE", "UNICODE_CI", "UTF8_BINARY")) {
```
Comment (mihailom-db) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala:623]: nit: I believe we used to test 4 collations everywhere, UTF8_BINARY, UTF8_LCASE, UNICODE and UNICODE_CI. These are kind of alternatives between spark implementations and ICU.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 10]
```diff
@@ -565,7 +565,15 @@ abstract class InterpretedHashFunction {
       case a: Array[Byte] =>
         hashUnsafeBytes(a, Platform.BYTE_ARRAY_OFFSET, a.length, seed)
       case s: UTF8String =>
-        hashUnsafeBytes(s.getBaseObject, s.getBaseOffset, s.numBytes(), seed)
+        val st = dataType.asInstanceOf[StringType]
+        if (st.supportsBinaryEquality) {
+          hashUnsafeBytes(s.getBaseObject, s.getBaseOffset, s.numBytes(), seed)
+        } else {
+          val stringHash = CollationFactory
+            .fetchCollation(dataType.asInstanceOf[StringType].collationId)
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:573]:
```suggestion
            .fetchCollation(st.collationId)
```

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM, with some nit comments

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**